File: creational-patterns/factory-method/demo-06/main.cc
#include <iostream>

#include "add_computing.h"
#include "computing_factory.h"
#include "minus_computing.h"

using namespace dp;

int main() {
  std::shared_ptr<AddComputing> add_ptr = ComputingFactory<AddComputing>::GetComputing(3, 4);
  int ret = add_ptr->DoComputing();
  std::cout << ret << std::endl;

  std::shared_ptr<MinusComputing> minus_ptr = ComputingFactory<MinusComputing>::GetComputing(3, 4);
  ret = minus_ptr->DoComputing();
  std::cout << ret << std::endl;

  return 0;
}

File: creational-patterns/factory-method/demo-06/main.lua
local ComputingFactory = require "computing_factory"
local SimpleComputingFactory = require "simple_computing_factory"

local AddComputing = require "add_computing"
local MinusComputing = require "minus_computing"

local function test_factory()
  local AddComputingFactory = ComputingFactory.New(AddComputing.name)
  local add_obj = AddComputingFactory:GetComputing(3, 4)
  print(add_obj:DoComputing())

  local MinusComputingFactory = ComputingFactory.New(MinusComputing.name)
  local add_obj = MinusComputingFactory:GetComputing(3, 4)
  print(add_obj:DoComputing())
end

local function test_simple_factory()
  local add_obj = SimpleComputingFactory.GetComputing(AddComputing.name, 3, 4)
  print(add_obj:DoComputing())

  local minus_obj = SimpleComputingFactory.GetComputing(MinusComputing.name, 3, 4)
  print(minus_obj:DoComputing())
end

test_factory()

test_simple_factory()
